File: sentinelx-ai/frontend/src/types/log.ts
/**
 * SentinelX AI – Log Collection Types
 * TypeScript interfaces matching the backend Pydantic v2 schemas for logs & log sources.
 */

export type LogSourceType =
  | "Syslog"
  | "Windows Event"
  | "Cloud Trail"
  | "Firewall"
  | "IDS/IPS"
  | "Endpoint"
  | "Application"
  | "Network"
  | "Other";

export type LogSourceStatus = "Active" | "Inactive" | "Error" | "Maintenance";

export type LogProtocol = "UDP" | "TCP" | "TLS" | "HTTPS" | "HTTP";

export type LogLevel = "TRACE" | "DEBUG" | "INFO" | "WARNING" | "ERROR" | "CRITICAL";

export interface LogSource {
  id: string;
  name: string;
  source_type: LogSourceType;
  vendor?: string | null;
  description?: string | null;
  hostname?: string | null;
  ip_address?: string | null;
  protocol?: LogProtocol | null;
  port?: number | null;
  status: LogSourceStatus;
  last_seen?: string | null;
  created_at: string;
  updated_at: string;
}

export interface LogSourceSummary {
  id: string;
  name: string;
  source_type: string;
  vendor?: string | null;
  hostname?: string | null;
  ip_address?: string | null;
  status: string;
  last_seen?: string | null;
  created_at: string;
  updated_at: string;
}

export interface LogSourceListResponse {
  total: number;
  page: number;
  page_size: number;
  items: LogSourceSummary[];
}

export interface LogEntry {
  id: string;
  source_id: string;
  asset_id?: string | null;
  event_timestamp: string;
  log_level: LogLevel;
  event_type: string;
  category?: string | null;
  message?: string | null;
  raw_log: Record<string, unknown>;
  source_ip?: string | null;
  destination_ip?: string | null;
  username?: string | null;
  process_name?: string | null;
  event_id?: string | null;
  correlation_id?: string | null;
  created_at: string;
}

export interface LogEntrySummary {
  id: string;
  source_id: string;
  asset_id?: string | null;
  event_timestamp: string;
  log_level: LogLevel | string;
  event_type: string;
  category?: string | null;
  message?: string | null;
  source_ip?: string | null;
  destination_ip?: string | null;
  username?: string | null;
  event_id?: string | null;
  created_at: string;
}

export interface LogEntryListResponse {
  total: number;
  page: number;
  page_size: number;
  items: LogEntrySummary[];
}

export interface LogEntryStatsResponse {
  total_entries: number;
  by_level: Record<string, number>;
  by_event_type: Record<string, number>;
  by_category: Record<string, number>;
}

export interface LogQueryParams {
  page?: number;
  page_size?: number;
  keyword?: string;
  level?: string;
  source_id?: string;
  asset_id?: string;
  username?: string;
  event_type?: string;
  category?: string;
  start_time?: string;
  end_time?: string;
}

export interface TopLogSourceItem {
  source_id: string;
  name: string;
  source_type: string;
  entry_count: number;
}

export interface LogVolumeBucket {
  time_bucket: string;
  count: number;
}

File: sentinelx-ai/backend/app/models/log.py
"""
SentinelX AI – Log Collection ORM Models
Schema: sentinelx.log_sources, sentinelx.log_entries
"""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any
from sqlalchemy import (
    String,
    Text,
    Integer,
    ForeignKey,
    DateTime,
    CheckConstraint,
    func,
)
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin

if TYPE_CHECKING:
    from app.models.asset import Asset


class LogSource(Base, TimestampMixin):
    """
    Log source model representing systems or services that emit security logs.
    Supports Syslog, Windows Event, Cloud Trail, Firewall, IDS/IPS, and more.
    """

    __tablename__ = "log_sources"
    __table_args__ = (
        CheckConstraint(
            "source_type IN ('Syslog', 'Windows Event', 'Cloud Trail', "
            "'Firewall', 'IDS/IPS', 'Endpoint', 'Application', 'Network', 'Other')",
            name="ck_log_sources_source_type",
        ),
        CheckConstraint(
            "status IN ('Active', 'Inactive', 'Error', 'Maintenance')",
            name="ck_log_sources_status",
        ),
        CheckConstraint(
            "protocol IS NULL OR protocol IN ('UDP', 'TCP', 'TLS', 'HTTPS', 'HTTP')",
            name="ck_log_sources_protocol",
        ),
        CheckConstraint(
            "port IS NULL OR (port >= 0 AND port <= 65535)",
            name="ck_log_sources_port",
        ),
        {"schema": "sentinelx"},
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )
    source_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        index=True,
    )
    vendor: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    hostname: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    ip_address: Mapped[str | None] = mapped_column(
        String(45),
        nullable=True,
    )
    protocol: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
    )
    port: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="Active",
        index=True,
    )
    last_seen: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    # Relationships
    entries: Mapped[list["LogEntry"]] = relationship(
        "LogEntry",
        back_populates="source",
        cascade="all, delete-orphan",
        lazy="select",
    )


class LogEntry(Base):
    """
    Log entry model representing individual security log events.
    Immutable records – no updated_at column; only created_at is tracked.
    """

    __tablename__ = "log_entries"
    __table_args__ = (
        CheckConstraint(
            "log_level IN ('TRACE', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL')",
            name="ck_log_entries_log_level",
        ),
        {"schema": "sentinelx"},
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    source_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sentinelx.log_sources.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    asset_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sentinelx.assets.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    event_timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    log_level: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="INFO",
        index=True,
    )
    event_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        index=True,
    )
    category: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )
    message: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )
    raw_log: Mapped[dict[str, Any]] = mapped_column(
        JSONB,
        nullable=False,
        default=dict,
        server_default="{}",
    )
    source_ip: Mapped[str | None] = mapped_column(
        String(45),
        nullable=True,
        index=True,
    )
    destination_ip: Mapped[str | None] = mapped_column(
        String(45),
        nullable=True,
        index=True,
    )
    username: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
        index=True,
    )
    process_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    event_id: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )
    correlation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )

    # Relationships
    source: Mapped["LogSource"] = relationship(
        "LogSource",
        back_populates="entries",
    )
    asset: Mapped["Asset | None"] = relationship(
        "Asset",
        foreign_keys=[asset_id],
        lazy="select",
    )
